base/compiler/inferencestate.jl

```diff
@@ -609,3 +629,228 @@ function narguments(sv::InferenceState, include_va::Bool=true)
     end
     return nargs
 end
+
+# IRInterpretationState
+# =====================
+
+# TODO add `result::InferenceResult` and put the irinterp result into the inference cache?
+struct IRInterpretationState
+    method_info::MethodInfo
+    ir::IRCode
+    mi::MethodInstance
+    world::UInt
+    curridx::RefValue{Int}
+    argtypes_refined::Vector{Bool}
+    sptypes::Vector{VarState}
+    tpdum::TwoPhaseDefUseMap
+    ssa_refined::BitSet
+    lazydomtree::LazyDomtree
+    valid_worlds::RefValue{WorldRange}
+    edges::Vector{Any}
+    parent # ::AbsIntState
+end
+
+# AbsIntState
+# ===========
+
+const AbsIntState = Union{InferenceState,IRInterpretationState}
```

Comment:
I made this `Union` rather than abstract type so that we can get better inferrability and allow JET to perform reasonable error analysis on `Core.Compiler`. I think there is no strong need to make `AbsIntState` as an extendable interface like `AbstractInterpreter`.